modules/internal/ArrayViewRankChange.chpl

```diff
@@ -22,9 +22,337 @@
+
+    // the higher-dimensional domain that we're equivalent to
+    var downdomPid:int;
+    var downdomInst: downdomtype(downrank, idxType, stridable);
+
+    //
+    // TODO: If we put this expression into the variable declaration
+    // above, we get a memory leak.  File a future against this?
+    //
+    proc downdomtype(param rank: int, type idxType, param stridable: bool) type {
+      var a = dist.downdist.newRectangularDom(rank=rank, idxType=idxType,
+                                              stridable=stridable);
+      return a.type;
+    }
+
+    proc downrank param {
+      return collapsedDim.size;
+    }
+
+    inline proc downdom {
+      if _isPrivatized(downdomInst) then
+        return chpl_getPrivatizedCopy(downdomInst.type, downdomPid);
+      else
+        return downdomInst;
+    }
+
+    proc dsiBuildArray(type eltType) {
+      pragma "no auto destroy"
+      const downarr = _newArray(downdom.dsiBuildArray(eltType));
+      return new ArrayViewRankChangeArr(eltType  =eltType,
+      // TODO: Update once we start privatizing vvv
+                                        _DomPid = nullPid,
+                                        dom = this,
+                                        _ArrPid=downarr._pid,
+                                        _ArrInstance=downarr._instance,
+                                        collapsedDim=collapsedDim,
+                                        idx=idx,
+                                        ownsArrInstance=true);
+    }
+
+    // TODO: Use delegation feature for these?
+    // TODO: Can't all these be implemented in ChapelArray given dsiDim()?
+
+    proc dsiNumIndices {
+      return updom.dsiNumIndices;
+    }
+
+    proc dsiLow {
+      return updom.dsiLow;
+    }
+
+    proc dsiHigh {
+      return updom.dsiHigh;
+    }
+
+    proc dsiAlignedLow {
+      return updom.dsiAlignedLow;
+    }
+
+    proc dsiAlignedHigh {
+      return updom.dsiAlignedHigh;
+    }
+
+    proc dsiStride {
+      return updom.dsiStride;
+    }
+
+    proc dsiAlignment {
+      return updom.dsiAlignment;
+    }
+
+    proc dsiFirst {
+      return updom.dsiFirst;
+    }
+
+    proc dsiLast {
+      return updom.dsiLast;
+    }
+
+    proc dsiDim(upDim: int) {
+      var downDim = 0;
+      for d in 1..downrank {
+        if !collapsedDim(d) {
+          downDim += 1;
+          if (upDim == downDim) then
+            return downdom.dsiDim(d);
+        }
+      }
+      halt("Fell out of loop in dsiDim():", (upDim, downDim));
+    }
+
+    proc dsiDims() {
+      var upDims: rank*range(idxType, stridable=stridable);
+      var upDim = 1;
+      for param downDim in 1..downrank {
+        if !collapsedDim(downDim) {
+          upDims(upDim) = downdom.dsiDim(downDim);
+          upDim += 1;
+        }
+      }
+      return upDims;
+    }
+
+    proc dsiGetIndices() {
+      return updom.dsiGetIndices();
+    }
+
+    proc dsiSetIndices(inds) {
```

Comment:
Ah, I see dsiNewRectangularDom in ArrayViewRankChangeDist calls it. Wouldn't it be better to name this method something else, so we can (potentially at least) have different behavior for assigning to a rank-change domain? Or at least to reduce confusion? E.g. finishInitializingWithIndices(inds).

Alternatively - maybe this code should be folded in to an alternative ArrayViewRankChangeDist constructor that accepts an 'inds' argument. But that might not go well until constructors/initializers are fixed.